README.md
# lisatools

Tools for monitoring my Lifetime ISA portfolio.

## Installation

```bash
$ pip install lisatools
```

## Usage

`lisatools` provides classes and functions that help me manage my Lifetime ISA
fund portfolio. For example, I can calculate the trades required to obtain
my target portfolio as in the following snippet:

```python
import lisatools
import datetime

ftse_global = lisatools.Fund(
    "FTSE Global All Cap Index Fund",
    172.14,
    isin="GB00BD3RZ582",
    date=datetime.date(2022, 11, 21)
)
gilts = lisatools.Fund(
    "U.K. Gilt UCITS ETF (VGOV)",
    18.58,
    isin="IE00B42WWV65",
    date=datetime.date(2022, 11, 21)
)
holding1 = lisatools.Holding(ftse_global, 1.0, 0.6)
holding2 = lisatools.Holding(gilts, 5.0, 0.4)
pf = lisatools.Portfolio([holding1, holding2])

buy, sell = pf.trade_to_target()
print("Buy:\n=====", buy, "\nSell:\n=====", sell, sep = "\n")
```

[A more elaborate example](/docs/example.ipynb) is provided in the /docs/ folder.

## License

`lisatools` was created by Istvan Kleijn. It is licensed under the terms of the MIT license.

## Credits

`lisatools` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.5.1 (2023-01-24)
### Fix
* **Portfolio.__getitem__:** Use operator.index ([`7402eca`](https://github.com/istvankleijn/lisatools/commit/7402eca4db71c4d946cc654a14a7ed4753f52b6f))

## v0.5.0 (2023-01-18)
### Feature
* Remember fund price data using time-to-live cache ([`d64c650`](https://github.com/istvankleijn/lisatools/commit/d64c65045257c6ec67b1cd7dd86e314e9a2471b8))
* **Portfolio.save:** Optionally silence output ([`458babb`](https://github.com/istvankleijn/lisatools/commit/458babb4f551f0bf93354e946632cb1088ac10b9))
* Provide command-line interface for price updates,  rebalancing, and new purchases ([`f43c4f3`](https://github.com/istvankleijn/lisatools/commit/f43c4f37d78dced00dc89a223b16234c56d91494))

## v0.4.0 (2023-01-04)
### Feature
* Extend JSON saving/loading to handle ETF holdings ([`d4a007a`](https://github.com/istvankleijn/lisatools/commit/d4a007aa9758d73ec800d218619c9b4c67481341))
* **Portfolio:** Add load and save functionality in JSON format ([`e2e383a`](https://github.com/istvankleijn/lisatools/commit/e2e383a707cc6b6a0fda3fa5ac23e52686c32b40))
* Provide custom JSON encoder and decoder ([`e2d69e5`](https://github.com/istvankleijn/lisatools/commit/e2d69e5d7448cbb0cf1c3b61e7488a4bfa4ae4ec))
* **Fund:** Use update_price to handle ISO strings ([`99e180e`](https://github.com/istvankleijn/lisatools/commit/99e180eaa89f937e799a5de89d8909404bc6a874))
* Convert Funds/Holdings to/from dictionaries ([`3055b6f`](https://github.com/istvankleijn/lisatools/commit/3055b6f377efe945229e9025bd03e2daa195eff3))

## v0.3.2 (2022-12-21)
### Fix
* Add dependencies for scraping ([`8b666ec`](https://github.com/istvankleijn/lisatools/commit/8b666ec98a284cd12c701fe25d447238878ae937))

## v0.3.1 (2022-12-21)
### Documentation
* Updated example notebook ([`8d88a84`](https://github.com/istvankleijn/lisatools/commit/8d88a84c645b2865962ace4c732d443c324db135))

## v0.3.0 (2022-12-02)
### Feature
* **Portfolio:** Construct portfolio directly from funds ([`98d3185`](https://github.com/istvankleijn/lisatools/commit/98d31854178d0a449a273132497a9fa55ea40778))
* **Portfolio:** Allow iteration over holdings ([`dcd11e3`](https://github.com/istvankleijn/lisatools/commit/dcd11e3ab1d55954ee7968a48c8e33f3dace3254))
* Implement equality checking ([`ffac284`](https://github.com/istvankleijn/lisatools/commit/ffac284ee32b6de956541cd8eb1011e3b30ee336))
* **Portfolio:** Implement basic sequence protocol ([`20f8b9a`](https://github.com/istvankleijn/lisatools/commit/20f8b9ac2eaaf889c3254eb7f55771b1948d3e21))
* Move Portfolio data to holdings attribute ([`da712e7`](https://github.com/istvankleijn/lisatools/commit/da712e7e3fd1cf2013c98b0049f38169fd4aaf16))

### Fix
* **Portfolio.from_funds:** Assert equal lengths of funds, units held, and target fractions ([`af543aa`](https://github.com/istvankleijn/lisatools/commit/af543aabd970136f160bdac32a695f25388bbc3e))

## v0.2.4 (2022-11-30)
### Fix
* Fix issues from initial flakeheaven run ([`41a29a5`](https://github.com/istvankleijn/lisatools/commit/41a29a546ea61f643fd4d51b07e9ad02031632da))

- Format code with Black
- Lint code with flakeheaven

## v0.2.3 (2022-11-29)

- Set up continuous deployment with PSR

## v0.2.1 (2022-11-29)

- Set up continuous integration
- Correct handling of current date

## v0.2.0 (2022-11-28)

- Added ETF class
- Implemented price updating using FT data scraping

## v0.1.0 (2022-11-23)

- First public release!
- Bugfixes
- Rewritten tests
- Expanded documentation

## v0.0.3 (2022-11-22)

- Implemented target portfolio and trades to reach it
- Added string representation methods
- Provided usage examples

## v0.0.2 (8/11/2022)

- Populated package namespace
- Added explicit implementation of `Fund` class

## v0.0.1 (1/11/2022)

- First release of `lisatools`!

src/lisatools/fund.py
import datetime


class Fund:
    """
    Details of a fund, including its current market price.

    Attributes
    ----------
    description : str
        Description of the fund, for example the name given by its provider.
    price : float, default 1.0
        Latest market price available for the fund.
    isin : str, default "None"
        If a real fund, the International Securities Identification Number
        (ISIN) of the fund should be specified.
        The ISIN is a unique 12-character alphanumerical identifier.
    date : datetime.date
        Date at which the `price` was last updated. If not specified, will be
        initialized to the current one (at runtime).

    Examples
    --------

    Constructing a Fund from a description and current price.

    >>> f = lisatools.Fund("FTSE Global All Cap Index Fund", 170.14)

    Constructing a Fund with all optional details.

    >>> f = lisatools.Fund("FTSE Global All Cap Index Fund", 170.14,
    ...                    isin="GB00BD3RZ582",
    ...                    date=datetime.date(2022, 11, 1))
    """

    def __init__(
        self,
        description="Default fund",
        price=1.0,
        *,
        isin="None",  # UNSPECIFIED9 is a valid ISIN
        date=None,
    ):
        self.description = description
        self.isin = isin
        self.update_price(price, date=date)

    def __repr__(self):
        return (
            f"Fund({self.description!r}, {self.price!r}, "
            f"date={self.date!r}, isin={self.isin!r})"
        )

    def __eq__(self, other):
        return (
            self.description == other.description
            and self.price == other.price
            and self.date == other.date
            and self.isin == other.isin
        )

    def update_price(self, price, *, date=None):
        """
        Set the price of the fund to a given value and optionally specify the
        date at which this price is correct.

        Parameters
        ----------
        price : float
            The new price of one unit of the fund.
        date : datetime.date, str, or None, default None
            The date when the fund had the given price. Can be provided as an
            ISO-formatted date string. If left unspecified, the date is set to the
            current one (at runtime).
        """
        self.price = price
        if isinstance(date, datetime.date):
            self.date = date
        elif isinstance(date, str):
            self.date = datetime.date.fromisoformat(date)
        elif date is None:
            self.date = datetime.date.today()
        else:
            raise TypeError(
                "date must be a `datetime.date` or an ISO-formatted date string"
            )

    def as_dict(self):
        """
        Encode the fund as a dictionary.

        The ISIN is encoded with dictionary key 'ISIN'; the other keys are equal to the
        names of the class attributes.
        """
        return {
            "ISIN": self.isin,
            "description": self.description,
            "price": self.price,
            "date": self.date,
        }

    @classmethod
    def from_dict(cls, d):
        """
        Construct a lisatools.Fund from a dictionary, respecting the class
        constructor defaults.

        The `isin` class attribute is set from the 'ISIN' dictionary key.

        Parameters
        ----------
        d : dict
            The dictionary from which the Fund is constructed.
        """
        return cls(
            d["description"],
            d["price"],
            isin=d.get("ISIN", "None"),
            date=d.get("date", None),
        )


class ETF(Fund):
    """
    Details of an exchange-traded fund, including its current market price.

    Attributes
    ----------
    description
        Description of the ETF. By default, this will be equal to the ticker
        symbol and the `name` attribute separated by a colon.
    price : float
        Latest market price available for the fund.
    name : str
        Name of the ETF, for example as given by its provider.
    isin : str
        If a real fund, the International Securities Identification Number
        (ISIN) of the fund should be specified.
        The ISIN is a unique 12-character alphanumerical identifier.
    date : datetime.date
        Date at which the `price` was last updated. If not specified, will be
        initialized to the current one (at runtime).

    Parameters
    ----------
    name : str
    price : float, default 1.0
    ticker : str or None, default None
    isin : str, default "None"
    date : datetime.date or None, default None

    Examples
    --------

    Constructing an ETF from a name and current price.

    >>> f = lisatools.ETF("U.K. Gilt UCITS ETF", 18.58)

    Constructing a Fund with all optional details.

    >>> f = lisatools.ETF(
    ...     "U.K. Gilt UCITS ETF",
    ...     18.58,
    ...     ticker="VGOV",
    ...     isin="IE00B42WWV65",
    ...     date=datetime.date(2022, 11, 21))
    """

    def __init__(self, name, price=1.0, *, ticker=None, isin="None", date=None):
        self.name = name
        if ticker is None:
            self.description = name
        else:
            self.description = f"{ticker}: {name}"
        self.ticker = ticker
        self.isin = isin
        self.update_price(price, date=date)

    def __repr__(self):
        return (
            f"ETF({self.name!r}, {self.price!r}, "
            f"date={self.date!r}, isin={self.isin!r}, ticker={self.ticker!r})"
        )

    def as_dict(self):
        """
        Encode the ETF as a dictionary.

        The ISIN is encoded with dictionary key 'ISIN'; the other keys are equal to the
        names of the class attributes.
        """
        return {
            "ISIN": self.isin,
            "description": self.description,
            "ticker": self.ticker,
            "name": self.name,
            "price": self.price,
            "date": self.date,
        }

    @classmethod
    def from_dict(cls, d):
        """
        Construct a lisatools.ETF from a dictionary, respecting the class
        constructor defaults.

        The `isin` class attribute is set from the 'ISIN' dictionary key.

        Parameters
        ----------
        d : dict
            The dictionary from which the ETF is constructed.
        """
        return cls(
            d["name"],
            d["price"],
            ticker=d.get("ticker", None),
            isin=d.get("ISIN", "None"),
            date=d.get("date", None),
        )
